File: Backend/src/main/java/com/tfg/truby_writer/model/services/ObjectsService.java
package com.tfg.truby_writer.model.services;

import java.nio.channels.OverlappingFileLockException;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Slice;
import org.springframework.stereotype.Service;
import jakarta.transaction.Transactional;
import com.tfg.truby_writer.model.entities.Character;

import com.tfg.truby_writer.model.enums.Enums;
import com.tfg.truby_writer.model.services.Block;
import com.tfg.truby_writer.model.entities.Event;
import com.tfg.truby_writer.model.entities.LineTime;

import com.tfg.truby_writer.model.exceptions.DuplicateInstanceException;
import com.tfg.truby_writer.model.exceptions.InstanceNotFoundException; 

import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.PageRequest;

public interface ObjectsService {

    //CHARACTERS

    Character addCharacter( Long projectId, String name, String bio, String mainImageUrl) throws DuplicateInstanceException, InstanceNotFoundException;

    Character modifyCharacter(Long characterId, String name, String bio, String mainImageUrl) throws InstanceNotFoundException;

    void deleteCharacter(Long characterId) throws InstanceNotFoundException;

    Character getCharacterById(Long characterId) throws InstanceNotFoundException;

    Block<Character> findCharactersByFilter(Long projectId, String text) throws InstanceNotFoundException;

    Character findCharacterByName(Long projectId, String name) throws InstanceNotFoundException;


    // EVENTS

    Event addEvent(Long timelineId, String name, String description, Integer chapterNumber, Integer chronoOrder) throws DuplicateInstanceException, InstanceNotFoundException;

    Event modifyEvent(Long id, String title, String content, Integer chapterNumber, Integer chronoOrder) throws InstanceNotFoundException, DuplicateInstanceException, IllegalArgumentException;

    void deleteEvent(Long eventId) throws InstanceNotFoundException;

    Block<Event> findEventsByName(Long timelineId, String title) throws InstanceNotFoundException;


    // LINE TIMES


    //FALTA HACER LA FUNCIÓN QUE ORDENE LOS EVENTOS DE UNA LINETIME POR ORDEN CRONOLOGICO O CAPITULAR

    LineTime getLineTimeByPlot(Long plotId) throws InstanceNotFoundException;
    
}

File: Frontend/src/backend/objectsService.js
import { appFetch } from './appFetch';

const BASE_PATH = '/structure';

export const addCharacter = async (projectId, characterDto) => {
    return await appFetch('POST', `${BASE_PATH}/projects/${projectId}/createCharacter`, characterDto);
};

export const modifyCharacter = async (characterId, characterDto) => {
    return await appFetch('PUT', `${BASE_PATH}/characters/modify/${characterId}`, characterDto);
};

export const deleteCharacter = async (characterId) => {
    return await appFetch('DELETE', `${BASE_PATH}/characters/delete/${characterId}`);
};

export const getCharacterById = async (characterId) => {
    return await appFetch('GET', `${BASE_PATH}/characters/${characterId}`);
};

export const findCharactersByFilter = async (projectId, text) => {
    return await appFetch('GET', `${BASE_PATH}/projects/${projectId}/characters/search?text=${encodeURIComponent(text)}`);
};

export const findCharacterByName = async (projectId, name) => {
    return await appFetch('GET', `${BASE_PATH}/projects/${projectId}/characters/byName?name=${encodeURIComponent(name)}`);
};

export const addEvent = async (timelineId, eventDto) => {
    return await appFetch('POST', `${BASE_PATH}/plots/${timelineId}/createEvent`, eventDto);
};

export const modifyEvent = async (eventId, eventDto) => {
    return await appFetch('PUT', `${BASE_PATH}/events/modify/${eventId}`, eventDto);
};

export const deleteEvent = async (eventId) => {
    return await appFetch('DELETE', `${BASE_PATH}/events/delete/${eventId}`);
};

export const findEventsByName = async (timelineId, title) => {
    return await appFetch('GET', `${BASE_PATH}/plots/${timelineId}/events/search?title=${encodeURIComponent(title)}`);
};

export const getLineTimeByPlot = async (plotId) => {
    return await appFetch('GET', `${BASE_PATH}/plots/${plotId}/lineTime`);
};